PHYSICS/AddRope.md

```diff
@@ -35,4 +85,46 @@ Rope will sometimes contract and fall to the ground like you'd expect it to, but
 * **unkPtr**: Unknown pointer, always 0 in original scrips.
 
 ## Return value
-A script handle for the rope
\ No newline at end of file
+A script handle for the rope
+
+## Examples
+```lua
+RegisterCommand("new_rope", function(soruce, args)
+
+    -- Get the handle for the player's ped
+    local pedHandle = PlayerPedId()
+
+    -- Ensure that the rope textures are loaded
+    local timer = 0
+    while not RopeAreTexturesLoaded() and timer < 1000 do
+        RopeLoadTextures()
+        timer = timer + 1
+        Citizen.Wait(0)
+    end
+
+    -- Get the coordinates for where the rope will be
+    local ropePos = GetOffsetFromEntityInWorldCoords(pedHandle, 0.0, 2.0, 0.5)
+
+    -- Create the rope
+    local newRopeHandle = AddRope(
+        ropePos.x, -- x
+        ropePos.y, -- y
+        ropePos.z, -- z
+        0.0, -- rotX
+        0.0, -- rotY
+        0.0, -- rotZ
+        10.0, -- maxLength
+        1, -- ropeType
+        10.0, -- initLength
+        0.0, -- minLength
+        1.0, -- lengthChangeRate
+        false, -- onlyPPU
+        false, -- collisionOn
+        false, -- lockFromFront
+        1.0, -- timeMultiplier
+        false, -- breakable
+        0 -- unkPtr
+    )

```

Comment:
Hmmm, why are we using a `timer` variable? This could cause the loop to break out if it ever goes over, which is unlikely, but still If this is going to take a while to load, it should be executed from a `thread` instead, i.e.

```lua
Citizen.CreateThread(function()
  -- Get the handle for the player's ped
  local pedHandle = PlayerPedId()
  -- Ensure that the rope textures are loaded
    
  while not RopeAreTexturesLoaded() do
    RopeLoadTextures()
  end
  
  -- Get the coordinates for where the rope will be
  local ropePos = GetOffsetFromEntityInWorldCoords(pedHandle, 0.0, 2.0, 0.5)
  -- Create the rope
  local newRopeHandle = AddRope(
      ropePos.x, -- x
      ropePos.y, -- y
      ropePos.z, -- z
      0.0, -- rotX
      0.0, -- rotY
      0.0, -- rotZ
      10.0, -- maxLength
      1, -- ropeType
      10.0, -- initLength
      0.0, -- minLength
      1.0, -- lengthChangeRate
      false, -- onlyPPU
      false, -- collisionOn
      false, -- lockFromFront
      1.0, -- timeMultiplier
      false, -- breakable
      0 -- unkPtr
  )
end)
```

That way it runs parallel to the main game thread.

Reply:
Thank you for the input, I totally agree with what you have said here and will make an updated example to reflect this better. I've always used the timer variable as a way to protect against getting stuck in a loop forever, this I typically used more with things such as loading models or animations where there is the possibility of a bad input or a valid input which isn't able to be loaded for some reason, however probably isn't as applicable with rope textures. 

I will touch up this example with your feedback in mind shortly.

Reply:
@colistro123 I have updated the example, it can be found [here](https://github.com/citizenfx/natives/pull/1145/files#diff-d378390d652ccf2c648ef059cf9b4b1b8fa20c619945d7bb502e047aba55d635R90-R137)

I moved the code into a thread as suggested and added a check after the while loop to ensure rope textures are loaded and check for that edge case. 

I opted to keep the timer and timeout for the while loop for the reason of preventing the very slim and odd edge case of the loop running forever, I don't see that ever happening however seeing as this is going in public documentation as an example I feel it's best to present code that doesn't have the potential of getting stuck in a infinite loop.

Reply:
I think that based on how the game internally loads those, that would be unlikely, but leave as is if you like. Point is that the game will always try to find an index to load those textures regardless.

![image](https://github.com/user-attachments/assets/d8c5f89f-b74e-4566-a1de-dd4bc9ae1e64)

Interestingly enough that function also gets called for specific vehicle fragment stuff, so it's unlikely for it to fail I suppose.

![image](https://github.com/user-attachments/assets/83b7f454-7999-4516-9368-241a934f8496)

Reply:
I see, that's some interesting info so thank you! I'll update the example accordingly